### DISCUSSION TITLE
Output of shear stress in CrystalPlasticitySlipRateGS

### DISCUSSION THREAD

**Discussion by Aayush-T** [idaholab/moose#21507] [Q&A Modules: Solid mechanics] [answered]
Hello everyone,
I am working on FiniteStrainUOBasedCP and CrystalPlasticitySlipRateGSS. I wanted the shear stress variable "tau" in CrystalPlasticitySlipRateGSS as output for all slip systems, so I tried to make tau from local to global variable, but it gives this error. 
**THE PROBLEM VARIABLE AND FUNCTION IS IN BOLD**
 Only showing the added portion in "CrystalPlasticitySlipRateGSS" file for better clarity and easy understanding.

Like this: 
```
.h file
#pragma once

#include "CrystalPlasticitySlipRate.h"
#include "RankTwoTensor.h"

/**
 * Phenomenological constitutive model slip rate userobject class.
 */
class CrystalPlasticitySlipRateGSS : public CrystalPlasticitySlipRate
{

protected:

  **DenseVector<Real> _shear_stress;**
};
```
and, in C file as
```

#include "CrystalPlasticitySlipRateGSS.h"

#include <fstream>

registerMooseObject("TensorMechanicsApp", CrystalPlasticitySlipRateGSS);

InputParameters
CrystalPlasticitySlipRateGSS::validParams()
{
.........
}

CrystalPlasticitySlipRateGSS::CrystalPlasticitySlipRateGSS(const InputParameters & parameters)
  : CrystalPlasticitySlipRate(parameters),
    .........,
    **_shear_stress(_variable_size)**
{
  if (_slip_sys_flow_prop_file_name.length() != 0)
    readFileFlowRateParams();
  else
    getFlowRateParams();
}

bool
CrystalPlasticitySlipRateGSS::calcSlipRate(unsigned int qp, Real dt, std::vector<Real> & val) const
{
  DenseVector<Real> tau(_variable_size);

  for (unsigned int i = 0; i < _variable_size; ++i)
    tau(i) = _pk2[qp].doubleContraction(_flow_direction[qp][i]);

  
   **for (unsigned int i = 0; i < _variable_size; ++i)**
     **_shear_stress(i) = _pk2[qp].doubleContraction(_flow_direction[qp][i]);**


  for (unsigned int i = 0; i < _variable_size; ++i)
  {
    val[i] = _a0(i) * std::pow(std::abs(tau(i) / _mat_prop_state_var[qp][i]), 1.0 / _xm(i)) *
             std::copysign(1.0, tau(i));
    if (std::abs(val[i] * dt) > _slip_incr_tol)
    {
#ifdef DEBUG
      mooseWarning("Maximum allowable slip increment exceeded ", std::abs(val[i]) * dt);
#endif
      return false;
    }
  }

  return true;
}
```

But it gives following error.
![error_file](https://user-images.githubusercontent.com/78248963/177389183-7aba3d41-e9a0-4c5e-8ca9-1e5fdfd7c5dc.png)

I am also adding the changes I made in the CrystalPlasticitySlipRateGSS 
![h file](https://user-images.githubusercontent.com/78248963/177389400-35332ffa-e172-4742-8f0c-c8d07dba8445.png)
![C file 1](https://user-images.githubusercontent.com/78248963/177389954-c8b31533-53cc-4c01-9119-a9af9c6d8a49.png)
![C file 2](https://user-images.githubusercontent.com/78248963/177389959-86b8bbec-dfc9-4611-8020-bc855c95b948.png)

**Comment by GiudGiud** [comment]
hello

You made calcSlipRate a `const` function but you are assigning a value to one of the class member variable.
This is not allowed, as declaring a function `const` means it may not modify any of the class attributes. 

You need to remove the const function attribute

Guillaume

**Reply by Aayush-T** [reply]
Hi, thanks for quick response. In this case, this is how this function should be defined(as the file is from MOOSEFRAMEWORK), otherwise it will lead to constitutive failure error in the file. I tried that.
I also tried a few more things such as trying to define a new function in the C file, but now is the problem of how to call that new function in the file. Moreover, I have doubts about how to get the output of a DenseVector in the csv when it is not defined as a material property.
```
#include "CrystalPlasticitySlipRateGSS.h"

#include <fstream>

registerMooseObject("TensorMechanicsApp", CrystalPlasticitySlipRateGSS);

InputParameters
CrystalPlasticitySlipRateGSS::validParams()
{
.......
}

CrystalPlasticitySlipRateGSS::CrystalPlasticitySlipRateGSS(const InputParameters & parameters)
  : CrystalPlasticitySlipRate(parameters),
    ...................,
    _shear_stress(_variable_size)
{
  if (_slip_sys_flow_prop_file_name.length() != 0)
    readFileFlowRateParams();
  else
    getFlowRateParams();
}

 **void
 CrystalPlasticitySlipRateGSS::CalcTau(unsigned int qp)
{
  for (unsigned int i = 0; i < _variable_size; ++i)
    _shear_stress(i) = _pk2[qp].doubleContraction(_flow_direction[qp][i]);
    printf("%f\n", _shear_stress(0));
}**

bool
CrystalPlasticitySlipRateGSS::calcSlipRate(unsigned int qp, Real dt, std::vector<Real> & val) const
{
  DenseVector<Real> tau(_variable_size);

  for (unsigned int i = 0; i < _variable_size; ++i)
    tau(i) = _pk2[qp].doubleContraction(_flow_direction[qp][i]);

  for (unsigned int i = 0; i < _variable_size; ++i)
  {
    val[i] = _a0(i) * std::pow(std::abs(tau(i) / _mat_prop_state_var[qp][i]), 1.0 / _xm(i)) *
             std::copysign(1.0, tau(i));
    if (std::abs(val[i] * dt) > _slip_incr_tol)
    {
#ifdef DEBUG
      mooseWarning("Maximum allowable slip increment exceeded ", std::abs(val[i]) * dt);
#endif
      return false;
    }
  }

  return true;
}
```
![Function_form](https://user-images.githubusercontent.com/78248963/177508233-f4f84127-a055-4a75-872c-0e26a9871b77.png)

**Reply by GiudGiud** [reply]
You ll need to mark`_shear_stress` as mutable if you want to modify it in a const function

**Comment by sapitts** [comment]
Hi @Aayush-T,

Have you considered using the material-based "Update" crystal plasticity classes (documentation page here: https://mooseframework.inl.gov/source/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.html ) instead of the older user object based system? The newer material-based system offers several advantages including significant simulation speedup, more robust substepping capabilities, and, cruicial for your application, the ability to easily output material properties like the shear stress and slip system strength. 

You can read more about the specific Kalidindi model you are using on the documentation page here: https://mooseframework.inl.gov/source/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.html. To then output the resolved shear stress in your output file, add the following to your input file:
```
[AuxVariables]
  [resolved_shear_stress_0]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[AuxKernels]
  [tau_0]
    type = MaterialStdVectorAux
    variable = resolved_shear_stress_0
    property = applied_shear_stress
    index = 0
    execute_on = timestep_end
  []
[]

[Postprocessors]
  [tau_0]
    type = ElementAverageValue
    variable = resolved_shear_stress_0
  []
[]
```
This example shows the necessary input file additions for the 0th slip system, and the pattern is the same for the other slip systems.

Hope this helps!

**Reply by Aayush-T** [reply]
Hi, thanks for the response. Yes, this new method seems better than older user object based system, but since most of development done by me is on that, so I wanted to do it in the UserObjectBased Class only. This discussion gave me a better understanding on how to implement it.
Thanks a lot.

**Comment by Aayush-T** [comment, marked as answer]
Hi everyone, thanks. I did it by defining rss as material property in FiniteStrainUOBasedCP and calculating it in calcResidual function in FiniteStrainUOBasedCP itself. The output can be taken in the same way as defined above.
It was an easy way to do it for user object based method.
Adding it here for further reference in case anyone needs it.

```
FiniteStrainUObasedCP::calcResidual()
{
RankTwoTensor iden(RankTwoTensor::initIdentity), ce, ee, ce_pk2, eqv_slip_incr, pk2_new;

getSlipRates();
if (_err_tol)
return;

for (unsigned int i = 0; i < _num_uo_slip_rates; ++i)
{
for (unsigned int j = 0; j < _uo_slip_rates[i]->variableSize(); ++j)
{
eqv_slip_incr += (_flow_direction[i])[_qp][j] * (_mat_prop_slip_rates[i])[_qp][j] * _dt;
_rss_output[_qp][j] = _pk2[_qp].doubleContraction((*_flow_direction[i])[_qp][j]);
}
}

eqv_slip_incr = iden - eqv_slip_incr;
_fp_inv = _fp_old_inv * eqv_slip_incr;
_fe = _dfgrd_tmp * _fp_inv;

ce = _fe.transpose() * _fe;
ee = ce - iden;
ee*= 0.5;

pk2_new = _elasticity_tensor[_qp] * ee;

_resid = _pk2[_qp] - pk2_new;
}
```
Thanks and Regards,
Aayush
